File: parser.js
#!/usr/bin/env node
/**
 * rom-meta-parser.js - Parse Android device tree metadata
 * Extracts ROM info from build.prop, device tree, fingerprint
 * Usage: ./parser.js --device pixel6 --prop ro.build.fingerprint
 */

const fs = require('fs');
const readline = require('readline');

// Common Android properties
const PROPS = {
  'ro.product.model': 'Device Model',
  'ro.build.version.release': 'Android Version',
  'ro.build.version.sdk': 'API Level',
  'ro.product.cpu.abi': 'CPU ABI',
  'ro.build.fingerprint': 'Build Fingerprint',
  'ro.build.id': 'Build ID',
  'ro.build.version.security_patch': 'Security Patch',
  'ro.bootloader': 'Bootloader',
  'ro.serialno': 'Serial Number',
  'ro.hardware': 'Hardware',
};

async function parseBuildProp(filepath) {
  const rl = readline.createInterface({
    input: fs.createReadStream(filepath),
    crlfDelay: Infinity
  });

  const props = {};
  for await (const line of rl) {
    const match = line.match(/^([^=]+)=(.+)$/);
    if (match) {
      props[match[1]] = match[2];
    }
  }
  return props;
}

function parseFingerprint(fingerprint) {
  // Format: brand/device/device:version/build_id/release/type/tags
  const parts = fingerprint.split('/');
  return {
    brand: parts[0],
    device: parts[1],
    codename: parts[2],
    version: parts[3],
    build_id: parts[4],
    release: parts[5],
    type: parts[6],
    tags: parts[7],
  };
}

async function main() {
  const args = process.argv.slice(2);
  
  if (args.includes('--help')) {
    console.log(`
rom-meta-parser - Parse Android device metadata

Usage:
  ./parser.js --file build.prop                  Parse build.prop file
  ./parser.js --fingerprint <fingerprint>       Parse fingerprint string
  ./parser.js --list                            List all properties

Examples:
  ./parser.js --file system/build.prop
  ./parser.js --fingerprint "google/oriole/oriole:13/TP1A.220624.014/8816326:user/release-keys"
    `);
    process.exit(0);
  }

  if (args.includes('--list')) {
    console.log('Known Android properties:\n');
    for (const [key, desc] of Object.entries(PROPS)) {
      console.log(`  ${key.padEnd(40)} ${desc}`);
    }
    return;
  }

  const fileIdx = args.indexOf('--file');
  if (fileIdx !== -1) {
    const filepath = args[fileIdx + 1];
    try {
      const props = await parseBuildProp(filepath);
      console.log(`Parsed ${Object.keys(props).length} properties from ${filepath}\n`);
      for (const [key, val] of Object.entries(props).slice(0, 20)) {
        const desc = PROPS[key] || '';
        console.log(`  ${key.padEnd(45)} = ${val}`);
      }
      if (Object.keys(props).length > 20) {
        console.log(`  ... and ${Object.keys(props).length - 20} more`);
      }
    } catch (e) {
      console.error(`Error reading file: ${e.message}`);
      process.exit(1);
    }
    return;
  }

  const fpIdx = args.indexOf('--fingerprint');
  if (fpIdx !== -1) {
    const fp = args[fpIdx + 1];
    const parsed = parseFingerprint(fp);
    console.log(`Fingerprint: ${fp}\n`);
    for (const [key, val] of Object.entries(parsed)) {
      console.log(`  ${key.padEnd(15)} ${val}`);
    }
    return;
  }

  console.log('Usage: ./parser.js --file build.prop');
  console.log('       ./parser.js --fingerprint <fingerprint>');
  console.log('       ./parser.js --help');
}

main().catch(console.error);

File: parser.py
#!/usr/bin/env python3
"""
parser.py -- Parse Android device tree metadata
Extracts: device specs, partition layout, build props, configs
Usage: python3 parser.py BoardConfig.mk
"""
import sys, re
from pathlib import Path

def parse_boardconfig(path):
    """Parse Android BoardConfig.mk"""
    content = Path(path).read_text()
    result = {}

    # Architecture
    arch_m = re.search(r'TARGET_ARCH\s*:=\s*(\w+)', content)
    if arch_m:
        result['arch'] = arch_m.group(1)

    # CPU variant
    cpu_m = re.search(r'TARGET_CPU_VARIANT\s*:=\s*(\S+)', content)
    if cpu_m:
        result['cpu'] = cpu_m.group(1)

    # Kernel
    kernel_m = re.search(r'TARGET_PREBUILT_KERNEL\s*:=\s*(\S+)', content)
    if kernel_m:
        result['kernel'] = kernel_m.group(1)

    # Partitions
    result['partitions'] = {}
    for line in content.splitlines():
        if 'BOARD_' in line and 'PARTITION' in line:
            key_m = re.search(r'(BOARD_\w+PARTITION\w+)\s*:=\s*(\S+)', line)
            if key_m:
                result['partitions'][key_m.group(1)] = key_m.group(2)

    # Recovery
    recovery_m = re.search(r'TARGET_RECOVERY_PIXEL_FORMAT\s*:=\s*(\w+)', content)
    if recovery_m:
        result['recovery_format'] = recovery_m.group(1)

    return result

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 parser.py <BoardConfig.mk>")
        sys.exit(1)

    result = parse_boardconfig(sys.argv[1])
    print("\n📖 Device Tree Metadata\n")
    for k, v in result.items():
        if isinstance(v, dict):
            print(f"{k}:")
            for sk, sv in v.items():
                print(f"  {sk:<40} {sv}")
        else:
            print(f"{k:<40} {v}")

if __name__ == "__main__":
    main()
